rfd/0135-kube-apps-discovery.md

```diff
@@ -0,0 +1,189 @@
+and update it on the backend, constantly keeping the dynamic apps state up to date. Writing to the backend will be handled
+by already existing mechanism in our code - reconciler. Reconciler will compare given fresh list of resources and resources we currently have
+on the backend, and then will decide on appropriate actions (create/update/delete) that are required on
+the backend to change resources to the desired state.
+The Teleport app service that runs in the same Kubernetes cluster will then react to those changes, and will start/stop proxying
+apps that are selected by appropriate labels.
+
+Name of the created Teleport app will consist of Kubernetes service name, namespace and cluster name, like this:
+`$SERVICE_NAME-$CLUSTER_NAME-$NAMESPACE`. This is done to avoid naming collisions in discovered apps. Though app names
+will be long in that case, there's ongoing work to improve UX when working with long resources names in Teleport, see [RFD 129](https://github.com/gravitational/teleport/pull/27258).
+
+To allow for finer control over exposed service transformation into Teleport app Kubernetes annotations will be used.
+By default we will be exposing services as `HTTP` apps, but annotation can be used to specify whether the service is a `TCP` application.
+
+If Kubernetes service has multiple ports in use, we will expose each port as a separate app, with name of the app including the port's name
+specified in the service. For example, if a service has port 443 exposed with name `tls` and also port `5432` with name `postgres`
+we will create two apps: `testApp-tls-main-cluster-prod`, `testApp-postgres-main-cluster-prod`. But user will be able to use annotation to specify
+preferred port, in that case only app with this port will be created, without port info in the app name.
+
+Labels from the Kubernetes service will be copied to the corresponding Teleport app.
+
+## UX
+
+### Discovery
+
+Discovery service will use new `kubernetes_apps` matchers to periodically poll list of services inside the Kubernetes cluster.
+Labels specified in the config will be used to filter out services that should be exposed. Also users will be able to
+specify namespaces to process.
+
+```yaml
+## This section configures the Discovery Service
+discovery_service:
+  enabled: yes
+  discovery_group: kube-auto-apps
+  kubernetes_apps:
+    - cluster_name: main-cluster
+      namespaces: [ "toronto", "porto" ]
+      labels:
+        env: staging
+    - cluster_name: main-cluster
+      namespaces: [ "seattle", "oakland" ]
+      labels:
+        env: testing
+```
+
+`cluster_name` field will be translated into `teleport.dev/kubeapp-cluster-name` label on the Teleport app resource, so
+later it's be easier to target it in the corresponding app service.
+
+Kubernetes annotation `teleport.dev/kubeapp-type` control type of the Teleport app we create. If annotation is missing
```

Comment:
Yep, I agree, the only thing I don't like is using very generic names. Probably something like `teleport.dev/discovery-type` and `teleport.dev/discovery-protocol` will be better. Also, not sure about mixing app and DB protocols in same annotations. 
Also thinking about mixing stuff, do we want to support exposing same kubernetes service as both app and DB? It could happen if service has more than one port, in that case separate annotations will allow expose it separately 🤔